File: backend/mallapi/src/main/java/org/nareun130/mallapi/domain/Product.java
package org.nareun130.mallapi.domain;

import java.util.ArrayList;
import java.util.List;

import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@Table(name = "tbl_product")
@Getter
@ToString(exclude = "imageList") //! imageList를 제외하고 가져온다.
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Product {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long pno;

    private String pname; 

    private int price;

    private String pdesc;

    private boolean delFlag;

    public void changeDel(boolean delFlag) {
        this.delFlag = delFlag;
    }

    @ElementCollection //* 값 타입 럴렉션 -> 기본적으로 lazy loading 방식으로 동작
    @Builder.Default
    private List<ProductImage> imageList = new ArrayList<>();

    public void changePrice(int price) {
        this.price = price;
    }

    public void changeDesc(String desc) {
        this.pdesc = desc;
    }

    public void changeName(String name) {
         this.pname = name;
    }

    public void addImage(ProductImage image) {
        image.setOrd(this.imageList.size());
        imageList.add(image);
    }

    public void addImageString(String fileName) {

        ProductImage productImage = ProductImage.builder()
            .fileName(fileName)
            .build();
            addImage(productImage);
    }

    public void clearList() {
        this.imageList.clear();;
    }

}

File: backend/mallapi-kotlin/src/main/kotlin/com/nareun/mallapikotlin/domain/Product.kt
package com.nareun.mallapikotlin.domain

import jakarta.persistence.*

@Entity
@Table(name = "tbl_product")
class Product(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var pno: Long = 0,

    var pname: String = "",

    var price: Int = 0,

    var pdesc: String = "",

    var delFlag: Boolean = false,

    @ElementCollection//컬렉션 객체임을 JPA가 알수 있도록, 엔티티가 x, 값타입 임베디드 타입에 대한 태이블 생성 & 1대다 관계로 다룸
    var imageList: MutableList<ProductImage> = mutableListOf()//ProductImage가 값 타입 객체이므로
) {

    fun changeDel(delFlag: Boolean) {
        this.delFlag = delFlag
    }

    fun changePrice(price: Int) {
        this.price = price
    }

    fun changeName(name: String) {
        this.pname = name
    }

    fun addImage(image: ProductImage) {
        image.ord = this.imageList.size
        imageList.add(image)
    }

    fun addImageString(fileName: String) {
        val productImage = ProductImage(fileName, this.imageList.size)
        addImage(productImage)
    }

    fun changeDesc(desc: String) {
        this.pdesc = desc
    }

    fun clearList() {
        this.imageList.clear()
    }

    override fun toString(): String {
        return "pno : ${this.pno} ,pname:${this.pname} ,price:${this.price} ,pdesc:${this.pdesc} ,delFlag:${this.delFlag}"
    }
}